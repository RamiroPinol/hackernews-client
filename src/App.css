body {
  margin: 0;
  color: #F3FFE2;
  background-color: #1D2127;
  background-image: url("noisy-texture.png");
  font: 400 14px CoreSans, Arial, sans-serif;
}

a {
  color: #F3FFE2;
}

a:hover {
  text-decoration: underline;
}

ul, li {
  list-style: none;
  padding: 0;
  margin: 0;
}

input {
  padding: 10px;
  border-radius: 5px;
  outline: none;
  margin-right: 10px;
  border: 1px solid #dddddd;
}

button {
  padding: 10px;
  border-radius: 5px;
  border: 1px solid #dddddd;
  background: transparent;
  color: #808080;
  cursor: pointer;
}

button:hover {
  color: #222;
}

*:focus {
  outline: none;
}

.page {
  margin: 20px;
}

header {
  background: #ff6600;
}

header h2 {
  margin: 0;
}

.table {
  margin: 20px 0;
}

.table-header {
  display: flex;
  line-height: 24px;
  font-size: 16px;
  padding: 0 10px;
  justify-content: space-between;
}

.table-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
}

.table-row {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  box-shadow: inset 0 2px 3px -2px rgba(255, 255, 255, 0.25), 0 -3px 2px -3px rgba(0, 0, 0, 0.75);
  border-top: 1px solid #0b0c0f;
}

.table-header>span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}

.table-row>span {
  overflow: hidden;
  text-overflow: ellipsis;
}

.button-inline {
  border-width: 0;
  background: transparent;
  color: inherit;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 0;
  font-size: inherit;
  cursor: pointer;
}

.button-active {
  border-radius: 0;
  border-bottom: 1px solid #38BB6C;
}

.active {
  background-color: red;
}

.active::after {
  content: " ˅";
}

.active.reverse::after {
  content: " ˄";
}

.item-title {
  width: 65%;
  padding: 0 5px;
}

.item-dismiss, {
  width: 10%;
}

.item-author  {
  width: 15%;
}

.item-comments, .item-points {
  width: 5%;
}
