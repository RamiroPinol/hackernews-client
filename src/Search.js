import React from 'react';
import PropTypes from 'prop-types';

function Search({ children, value, onChange, onSubmit }) {
  return (
    <form onSubmit={onSubmit}>
      <input
        type="text"
        value={value}
        onChange={onChange}
      />
      <button type="submit">
        {children}
      </button>
    </form>
  );
}

Search.propTypes = {
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  value: PropTypes.string.isRequired,
};

export default Search;
