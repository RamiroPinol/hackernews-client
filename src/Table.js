import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';

function Table({ list, onDismiss }) {
  return (
    <div className="table">
      {list.map(item =>
        <div key={item.objectID} className="table-row">
          <span>
            <a href={item.url}>{item.title}</a>
          </span>
          <span>{item.author}</span>
          <span>{item.num_comments}</span>
          <span>{item.points}</span>
          <span>
            <Button
              className="button-inline"
              onClick={() => onDismiss(item.objectID)}
            >
              Dismiss
            </Button>
          </span>
        </div>)}
    </div>
  );
}

Table.propTypes = {
  onDismiss: PropTypes.func.isRequired,
};

export default Table;
