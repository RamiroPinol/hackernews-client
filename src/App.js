import React, { Component } from 'react';
import Search from './Search';
import Table from './Table';
import Button from './Button';
import Loading from './Loading';
import './App.css';

const DEFAULT_QUERY = 'redux';
const DEFAULT_PAGE = 0;
const DEFAULT_HPP = '100';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchTerm: DEFAULT_QUERY,
      searchKey: '',
      isLoading: false,
    };

    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.setSearchTopstories = this.setSearchTopstories.bind(this);
    this.fetchSearchTopstories = this.fetchSearchTopstories.bind(this);
    this.needsToSearchTopstories = this.needsToSearchTopstories.bind(this);
  }

  setSearchTopstories(result) {
    const { hits, page } = result;
    const { searchKey, results } = this.state;

    const oldHits = results && results[searchKey] ? results[searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];
    this.setState({
      results: { ...results, [searchKey]: { hits: updatedHits, page } },
      isLoading: false,
    });
  }

  needsToSearchTopstories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  fetchSearchTopstories(searchTerm, page) {
    this.setState({ isLoading: true });

    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}
      &${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(res => res.json())
      .then(result => this.setSearchTopstories(result));
  }

  componentDidMount() {
    this.setState({ searchKey: this.state.searchTerm });
    this.fetchSearchTopstories(this.state.searchTerm, DEFAULT_PAGE);
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const updatedHits = hits.filter(item => item.objectID !== id);
    this.setState({ results: { ...results, [searchKey]: { hits: updatedHits, page } } });
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  onSearchSubmit(event) {
    this.setState({ searchKey: this.state.searchTerm });
    if (this.needsToSearchTopstories(this.state.searchTerm)) {
      this.fetchSearchTopstories(this.state.searchTerm, DEFAULT_PAGE);
    }
    event.preventDefault();
  }

  render() {
    const { searchTerm, results, searchKey, isLoading } = this.state;
    const page = (results && results[searchKey] && results[searchKey].page) || 0;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="page">
        <div className="interactions">
          <div className="App">
            <Search
              value={searchTerm}
              onChange={this.onSearchChange}
              onSubmit={this.onSearchSubmit}
            >Search</Search>
            <Table
              list={list}
              onDismiss={this.onDismiss}
            />
            <div className="interactions">
              {isLoading ?
                <Loading /> :
                <Button onClick={() => this.fetchSearchTopstories(searchKey, page + 1)}>
                  More
                </Button>
              }
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default App;
