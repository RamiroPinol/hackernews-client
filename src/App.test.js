import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme'
import App from './App';
import Search from './Search';
import Button from './Button';
import Table from './Table';

// App Tests
describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });

  test('snapshots', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// Search Tests
describe('Search', () => {
  it('renders correctly', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Search</Search>, div);
  });

  test('snapshots', () => {
    const component = renderer.create(<Search>Search</Search>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// Button Tests
describe('Button', () => {
  it('renders correctly', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Test button</Button>, div);
  });

  test('snapshots', () => {
    const component = renderer.create(<Button>Test button</Button>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// Table Tests
describe('Table', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
    ],
  };

  it('renders correctly', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('snapshots', () => {
    const component = renderer.create(<Table {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items in list', () => {
    const element = shallow(<Table {...props} />);
    expect(element.find('.table-row').length).toBe(2);
  });
});
