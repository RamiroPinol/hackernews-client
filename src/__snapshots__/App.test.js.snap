exports[`App snapshots 1`] = `
<div
  className="page">
  <div
    className="interactions">
    <div
      className="App">
      <form
        onSubmit={[Function]}>
        <input
          onChange={[Function]}
          type="text"
          value="redux" />
        <button
          type="submit">
          Search
        </button>
      </form>
      <div
        className="table" />
      <div
        className="interactions">
        <div>
          Loading ...
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Button snapshots 1`] = `
<button
  className=""
  onClick={undefined}
  type="button">
  Test button
</button>
`;

exports[`Search snapshots 1`] = `
<form
  onSubmit={undefined}>
  <input
    onChange={undefined}
    type="text"
    value={undefined} />
  <button
    type="submit">
    Search
  </button>
</form>
`;

exports[`Table snapshots 1`] = `
<div
  className="table">
  <div
    className="table-row">
    <span>
      <a
        href={undefined}>
        1
      </a>
    </span>
    <span>
      1
    </span>
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row">
    <span>
      <a
        href={undefined}>
        2
      </a>
    </span>
    <span>
      2
    </span>
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
